packages/core/src/common/message-service.ts

```diff
@@ -6,29 +6,71 @@
+     * The timeout value is set in milisec, so 5000 = 5000 msec = 5 sec
+     * If we set {timeout: 0 }, then the notification preference will be ignored and notification
+     * remains until user intervention.
+     */
 
-    log(message: string, ...actions: string[]): Promise<string | undefined> {
-        return this.client.showMessage({ type: MessageType.Log, text: message, actions: actions || [] });
+    log(message: string, ...actions: string[]): Promise<string | undefined>;
+    log(message: string, options?: MessageOptions): Promise<string | undefined>;
+    log(message: string, options?: MessageOptions, ...actions: string[]): Promise<string | undefined>;
+    log(message: string, ...args: any[]): Promise<string | undefined> {
+        return this.processArgs(MessageType.Log, message, args);
     }
 
-    info(message: string, ...actions: string[]): Promise<string | undefined> {
-        return this.client.showMessage({ type: MessageType.Info, text: message, actions: actions || [] });
+    info(message: string, ...actions: string[]): Promise<string | undefined>;
+    info(message: string, options?: MessageOptions): Promise<string | undefined>;
+    info(message: string, options?: MessageOptions, ...actions: string[]): Promise<string | undefined>;
+    info(message: string, ...args: any[]): Promise<string | undefined> {
+        return this.processArgs(MessageType.Info, message, args);
     }
 
-    warn(message: string, ...actions: string[]): Promise<string | undefined> {
-        return this.client.showMessage({ type: MessageType.Warning, text: message, actions: actions || [] });
+    warn(message: string, ...actions: string[]): Promise<string | undefined>;
+    warn(message: string, options?: MessageOptions): Promise<string | undefined>;
+    warn(message: string, options?: MessageOptions, ...actions: string[]): Promise<string | undefined>;
+    warn(message: string, ...args: any[]): Promise<string | undefined> {
+        return this.processArgs(MessageType.Warning, message, args);
     }
 
-    error(message: string, ...actions: string[]): Promise<string | undefined> {
-        return this.client.showMessage({ type: MessageType.Error, text: message, actions: actions || [] });
+    error(message: string, ...actions: string[]): Promise<string | undefined>;
+    error(message: string, options?: MessageOptions): Promise<string | undefined>;
+    error(message: string, options?: MessageOptions, ...actions: string[]): Promise<string | undefined>;
+
+    error(message: string, ...args: any[]): Promise<string | undefined> {
+        return this.processArgs(MessageType.Error, message, args);
+    }
+
+    private processArgs(messageType: MessageType, messageText: string, args: any[]): Promise<string | undefined> {
+        let timeoutvalue: number | undefined;
+        let actionsValue: string[] | undefined;
+
+        if (args && args.length > 0) {
+            if (args[0]['timeout'] !== undefined) {
```

Comment:
please use [type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html) if possible. also, I think we can assume `args[0]` to be message options if type is object.

Reply:
would be nice, if you add the type guard in the namespace of `MessageOptions`.

Reply:
In the next version, I remove theoverloa method by addibg the actions and timeout within the same API, so the typeguard are not require, the code is simplier here